Search feature phase 1 implementation

seav opened the pull request:
### Proposed changes in this pull request
- Add basic full-text search functionality to the platform using an Elasticsearch infrastructure:
    - Add a new search page view where users can perform searches and see results.
    - Add a new "Search" tab to the project sidebar that shows a search box when clicked and goes to the search page when submitted.
    - Add a new `search` Django app to contain the search-related backend code.
    - Add a new `search` namespace to the UI and API URL config.
    - Add a `ui_class_name` and `ui_detail_url` to the `Resource` model to match the other record models. This is to facilitate uniform display of search results across entity types.
    - Add a very simple default view to serve the search page.
    - Add a search API endpoint that is accessed by the search page via AJAX. The API then translates the UI query into an ES query and then queries the ES API. The ES response is then parsed and translated into HTML and returned to the search page for display as search results.
    - Update the Squid config to prevent it from caching ES API responses.
    - Add the requests Python package to the common requirements.

### When should this PR be merged
When convenient.

### Risks
None foreseen.

### Follow up actions
- When deploying, install the **requests** Python package.
- When deploying, the `ES_*` settings should be set to the actual ES API endpoint for the corresponding site.


### Checklist (for reviewing)

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
- [ ] **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
- [ ] **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts. 

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
- [ ] **If this is a bug fix, are tests for the issue in place**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.


#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

bjohare commented on cadasta/templates/organization/project_wrapper.html:
```diff
@@ -134,6 +140,39 @@ <h1 class="{% if is_administrator or is_allowed_add_location or is_allowed_add_r
+'use strict';
+window.addEventListener('load', function() {
+  $('#sidebar .nav-sidebar .search a').bind('click', function(e) {
+    console.log('print')
```
This could be removed..

seav replied:
Yes, I'll remove this. This was also mentioned by @bjohare above.

seav replied:
Yes, I'll remove this.

bjohare commented on cadasta/resources/models.py:
```diff
@@ -122,6 +123,21 @@ def save(self, *args, **kwargs):
         create_thumbnails(self, (not self.id))
         super().save(*args, **kwargs)
 
+    @property
```
Is there a way to calculate these properties in view code rather than adding UI related stuff to the models?

oliverroick replied:
That's actually good practice. _Skinny views, fat models_ they say

bjohare replied:
Meh.. feels wrong..

seav replied:
Well, I was just following the pattern established last year in [this discussion](https://github.com/Cadasta/cadasta-platform/pull/644#discussion-diff-77601972). It certainly simplifies the view and adds easy to test methods on the model.

oliverroick commented on cadasta/search/views/api.py:
```diff
@@ -0,0 +1,199 @@
+    def get_main_label(self, model, source):
+        """Returns the search result's main UI label."""
+        if model == SpatialUnit:
+            for key, item in SPATIAL_TYPE_CHOICES:
```
Instead of looping over all choices you could convert `SPATIAL_TYPE_CHOICES` into a `dict` after it is imported and you can query the `dict` here:

```python
SPATIAL_TYPE_CHOICES = {c[0]: c[1] for c in SPATIAL_TYPE_CHOICES}

...

if model == SpatialUnit:
    return SPATIAL_TYPE_CHOICES.get(source['type'], '-')
```

seav replied:
OK, I'll convert this into a `dict`.

oliverroick commented on cadasta/search/views/api.py:
```diff
@@ -0,0 +1,199 @@
+            ])
+        elif type(entity) == Party:
+            label = '—'
+            for key, item in Party.TYPE_CHOICES:
```
I would convert the `tuple` into a `dict` again and query the `dict`. Or at least add a `break` statement to the loop so execution of the loop stops when the label is found.

seav replied:
OK, I'll create a `PARTY_TYPE_CHOICES` `dict` for this, similar to the `SPATIAL_TYPE_CHOICES` suggestion above.

bjohare commented on cadasta/search/views/api.py:
```diff
@@ -0,0 +1,199 @@
+api = settings.ES_SCHEME + '://' + settings.ES_HOST + ':' + settings.ES_PORT
+
+
+class Search(APIPermissionRequiredMixin,
```
This all fits on one line

seav replied:
OK. I'll reformat the code.

bjohare commented on cadasta/search/views/api.py:
```diff
@@ -0,0 +1,199 @@
+    def get_entity(self, es_type, source):
+        """Returns the model instance for a search result given its ES type and
+        the result source document, which should contain the database ID."""
+        mappings = (
```
This method might be more readable if `mappings` was implemented as a dict?

seav replied:
Yes, I think so. I can't remember why I structured it this way. I'll change this into a `dict`.

seav replied:
Yes. (Also mentioned by @oliverroick.)

bjohare commented on cadasta/search/views/default.py:
```diff
@@ -0,0 +1,23 @@
+from organization.views.mixins import ProjectMixin
+
+
+class Search(LoginPermissionRequiredMixin,
```
This fits on one line.

seav replied:
OK. I'll reformat the code.

bjohare commented on cadasta/templates/search/search.html:
```diff
@@ -0,0 +1,153 @@
+                  {% blocktrans %}
+                  <ul>
+                    <li>Review your wording, make sure that there aren’t any typos in your search query</li>
+                    <li>If you are having no results matching your query, try removing some search terms (“John”, instead of “John White”)</li>
```
This is expressed a little awkwardly.. maybe something like: 'If no results match your query,..'

seav replied:
Yeah, it is awkward. I'll improve this.

seav commented on cadasta/config/settings/default.py:
```diff
@@ -522,3 +518,7 @@
     'xls': 'organization.importers.xls.XLSImporter'
 }
+
+ES_SCHEME = 'http'
```
I think this is needed in case a deployment (maybe not in our case) wants encrypted API accesses via HTTPS.

seav commented on cadasta/core/static/css/single.scss:
```diff
@@ -277,22 +283,62 @@
+  z-index: 1000;
+  background: $brand-darkblue;
+  overflow: hidden;
+  transition: width 0.5s;
```
Makes sense. OK, I'll make this a slider.

seav commented on cadasta/resources/models.py:
```diff
@@ -122,6 +123,21 @@ def save(self, *args, **kwargs):
+        return _("Resource")
+
+    @property
+    def ui_detail_url(self):
```
I see. It should be simple to just rename all methods to `get_absolute_url` and also include the other recommendations for this method (such as properly URL-encoding characters).

seav commented on cadasta/search/views/api.py:
```diff
@@ -0,0 +1,199 @@
+             ProjectMixin,
+             APIView):
+
+    permission_required = 'project.view_private'
```
This is because **all** users, including not-logged-in users have `project.view` permissions on all public projects. See this [policy file](https://github.com/Cadasta/cadasta-platform/blob/master/cadasta/config/permissions/default.json#L27). I'm using `project.view_private` as a reasonable proxy search permission in lieu of checking that the user has _any_ of the `spatial.list`, `party.list`, `tenure_rel.list`, or `resource.list` permissions.

seav commented on cadasta/search/views/default.py:
```diff
@@ -0,0 +1,23 @@
+             ProjectMixin,
+             TemplateView):
+    template_name = 'search/search.html'
+    permission_required = 'project.view_private'
```
See reply on a similar comment above.

seav commented on cadasta/search/views/api.py:
```diff
@@ -0,0 +1,199 @@
+import requests
```
Yeah, I was actually wondering if this is a real API. OK. I'll convert this into an `async` module instead. I think in the future we can provide a _real_ API endpoint and refactor common code into a mixin. This API would be useful for apps like the QGIS plugin.

seav commented on cadasta/search/views/api.py:
```diff
@@ -0,0 +1,199 @@
+
+    def htmlize_result(self, result):
+        """Formats the search result into an HTML snippet."""
+        html = (
```
Yes, that is reasonable. I'll make a template; it should make @clash99's work easier. :)

seav commented:
> We should try to return results when working on the dev VM. I actually thought that this work was part of #910. Having example search results in place is important because it will help tweak the UI if necessary or optimise workflows without having to edit on staging, which is tedious.

I'll see what I can do.

> Do we need to add search to the API? It's probably a separate task, which we can consider implementing at some later point.

See my reply on the [`async` feedback](https://github.com/Cadasta/cadasta-platform/pull/1075#discussion_r97741016).

seav commented:
@oliverroick, @bjohare: I've finished responding to all comments and updated the code for those feedback items that I had committed to correct.

seav commented:
Rebased now.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Cadasta/cadasta-platform?pullRequest=1075) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Cadasta/cadasta-platform?pullRequest=1075) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: seav<br/>:x: clash99

seav commented:
@oliverroick, @bjohare: Aside from the UI/CSS tweaks added by @clash99, I've added a couple of updated functionality in my latest commit:

1. Before, if the ES server is unavailable or an error somehow occurred (the ES' HTTP response is not 200 OK), I raise an assertion. This unfortunately leaves the user with a loading GIF until the AJAX query times out. Now I added an explicit "search index is unavailable" message so the user is not left hanging.
2. On the front-end, I added search results persistence so that if the user refreshes the search page, the results are shown again instead of being provided with a blank form and results. Basically, the browser location is always updated with the last-performed query.
